File: frontend/src/lib/api/ingest_sources.ts
import { api } from "./client";

export interface SourceCandidate {
  source_kind: string;
  external_id: string;
  title: string;
  uploader: string | null;
  duration_seconds: number | null;
  thumbnail_url: string | null;
  webpage_url: string;
  already_in_library: boolean;
}

export interface SourceState {
  kind: string;
  name: string;
  enabled: boolean;
  state: string;
  consecutive_failures: number;
  last_error: string | null;
  rate_limit_max: number;
  rate_limit_window_seconds: number;
}

export interface SourceSearchResult {
  items: SourceCandidate[];
  errored_sources: string[];
}

export async function searchSources(q: string): Promise<SourceSearchResult> {
  return api<SourceSearchResult>("/api/ingest-sources/search", {
    method: "POST",
    body: JSON.stringify({ q }),
  });
}

export async function listSources(): Promise<SourceState[]> {
  const { items } = await api<{ items: SourceState[] }>(
    "/api/admin/ingest-sources",
  );
  return items;
}

export const setSourceEnabled = (kind: string, enabled: boolean) =>
  api<SourceState>(`/api/admin/ingest-sources/${encodeURIComponent(kind)}`, {
    method: "PUT",
    body: JSON.stringify({ enabled }),
  });

export const setSourceRateLimit = (
  kind: string,
  enabled: boolean,
  rate_limit_max: number,
  rate_limit_window_seconds: number,
) =>
  api<SourceState>(`/api/admin/ingest-sources/${encodeURIComponent(kind)}`, {
    method: "PUT",
    body: JSON.stringify({ enabled, rate_limit_max, rate_limit_window_seconds }),
  });

File: apps/web/src/setvault_web/api/ingest_sources.py
from __future__ import annotations

import logging
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from setvault_core.models.url_rip import RipJob
from setvault_core.services.ingest_sources import (
    list_states,
    search_all_sources,
    set_enabled,
)
from setvault_ingest_sources.base import Candidate
from setvault_ingest_sources.registry import get_source
from sqlalchemy import false, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from setvault_web.deps import current_user, db_session, require_admin
from setvault_web.schemas.ingest_sources import (
    CandidateOut,
    SetEnabledIn,
    SourceSearchIn,
    SourceSearchOut,
    SourceStateOut,
    SourceStatesOut,
)

logger = logging.getLogger(__name__)

router = APIRouter(tags=["ingest-sources"])


def _state_out(st, source) -> SourceStateOut:
    return SourceStateOut(
        kind=st.kind,
        name=(source.name if source else st.kind),
        enabled=st.enabled,
        state=st.state,
        consecutive_failures=st.consecutive_failures,
        last_error=st.last_error,
        rate_limit_max=st.rate_limit_max,
        rate_limit_window_seconds=st.rate_limit_window_seconds,
    )


@router.post("/api/ingest-sources/search", response_model=SourceSearchOut)
async def search(
    body: SourceSearchIn,
    user: Annotated[object, Depends(current_user)],
    session: Annotated[AsyncSession, Depends(db_session)],
) -> SourceSearchOut:
    result = await search_all_sources(
        session, query=body.q, limit_per_source=body.limit_per_source
    )
    cands = result.candidates

    id_keys: set[tuple[str, str]] = set()
    urls: set[str] = set()
    if cands:
        platforms = {c.source_kind for c in cands}
        ext_ids = [c.external_id for c in cands if c.external_id]
        webpage_urls = [c.webpage_url for c in cands if c.webpage_url]
        # Scope to the candidate keys so this stays cheap on a large library
        # rather than scanning every non-failed rip for the searched platforms.
        rows = (await session.execute(
            select(RipJob.source_platform, RipJob.source_external_id, RipJob.source_url).where(
                RipJob.source_platform.in_(platforms),
                RipJob.status != "failed",
                or_(
                    RipJob.source_external_id.in_(ext_ids) if ext_ids else false(),
                    RipJob.source_url.in_(webpage_urls) if webpage_urls else false(),
                ),
            )
        )).all()
        for platform, ext_id, src_url in rows:
            if ext_id:
                id_keys.add((platform, ext_id))
            if src_url:
                urls.add(src_url)

    def _in_lib(c: Candidate) -> bool:
        # URL fallback (for sources with no stable external_id) is an exact-string
        # match, so trailing-slash/query drift can miss — chromaprint is the real dedup.
        return (c.source_kind, c.external_id) in id_keys or c.webpage_url in urls

    return SourceSearchOut(
        items=[
            CandidateOut(
                source_kind=c.source_kind, external_id=c.external_id, title=c.title,
                uploader=c.uploader, duration_seconds=c.duration_seconds,
                thumbnail_url=c.thumbnail_url, webpage_url=c.webpage_url,
                already_in_library=_in_lib(c),
            )
            for c in cands
        ],
        errored_sources=result.errored_kinds,
    )


@router.get("/api/admin/ingest-sources", response_model=SourceStatesOut)
async def admin_list(
    _: Annotated[object, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(db_session)],
) -> SourceStatesOut:
    states = await list_states(session)
    return SourceStatesOut(items=[
        _state_out(s, get_source(s.kind)) for s in states
    ])


@router.put("/api/admin/ingest-sources/{kind}", response_model=SourceStateOut)
async def admin_set_enabled(
    kind: str,
    body: SetEnabledIn,
    _: Annotated[object, Depends(require_admin)],
    session: Annotated[AsyncSession, Depends(db_session)],
) -> SourceStateOut:
    source = get_source(kind)
    if source is None:
        raise HTTPException(status_code=404, detail="unknown source")
    st = await set_enabled(session, kind, body.enabled)
    if body.rate_limit_max is not None:
        st.rate_limit_max = body.rate_limit_max
    if body.rate_limit_window_seconds is not None:
        st.rate_limit_window_seconds = body.rate_limit_window_seconds
    await session.flush()
    await session.commit()
    return _state_out(st, source)
